designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,349 @@
+      1. [List Approach: Multiple Reasons per Budget - Recommended](#list-approach-multiple-reasons-per-budget---recommended)
+      2. [Single Reason Approach: One Reason per Budget](#single-reason-approach-one-reason-per-budget)
+   2. [Approach B: Defining Per Reason Controls](#approach-b-defining-per-reason-controls)
+   3. [Preferred Option: List Approach](#preferred-option-list-approach)
+   4. [Disruption Behavior](#disruption-behavior)
+   5. [Scenarios](#scenarios)
+      1. [Block Drift and Allow Expiration](#block-drift-and-allow-expiration)
+      2. [Fully Blocking Drift & Expiration while Allowing Consolidation](#fully-blocking-drift--expiration-while-allowing-consolidation)
+      3. [Fully Blocking Drift and Expiration while Allowing Emptiness](#fully-blocking-drift-and-expiration-while-allowing-emptiness)
+      4. [Limiting Drift & Expiration while Allowing Aggressive but Not Unbounded Consolidation](#limiting-drift--expiration-while-allowing-aggressive-but-not-unbounded-consolidation)
+      5. [Limiting Cost Saving Disruption to Occur During Weekends](#limiting-cost-saving-disruption-to-occur-during-weekends)
+      6. [How Should We Handle an Unspecified Reason When Others are Specified?](#how-should-we-handle-an-unspecified-reason-when-others-are-specified)
+
+## User Scenarios 
+1. Users need the capability to schedule upgrades only during business hours or within more restricted time windows. Additionally, they require a system that doesn't compromise the cost savings from consolidation when upgrades are blocked due to drift.
+2. Users want to minimize workload disruptions during business hours but still want to be able to delete empty nodes throughout the day.  That is, empty can run all day, while limiting cost savings and upgrades due to drift to non-business hours only.
+
+See Less Made Up Scenarios here: 
+- https://github.com/kubernetes-sigs/karpenter/issues/924 
+- https://github.com/kubernetes-sigs/karpenter/issues/753#issuecomment-1790110838
+- https://github.com/kubernetes-sigs/karpenter/issues/672
+
+See further breakdown of some additional scenarios towards the bottom of the document 
+
+## Clarifying the requirements and behavior 
+**Reason and Budget Definition:** Users should be able to define an reason and a corresponding budget(s).
+**Supported Reasons:** All disruption Reasons affected by the current Budgets implementation (underutilized, empty, expired, drifted) should be supported. 
+**Default Behavior for Unspecified Reasons:** Budgets should continue to support a default behavior for all disruption reasons. 
+
+
+# API Design
+## Approach A: Extending the Budget API to specify a reason 
+This section contrasts two approaches for specifying disruption reasons in the v1beta1 nodepool API: a list of reasons (List Approach) and a single reason per budget (Single Reason Approach).
+
+### List Approach: Multiple Reasons per Budget - Recommended
```

Comment:
@Bryce-Soghigian as this solution is recommended I would like to ask if we can extend this approach with `spec.disruption.defaults.budgets` section as proposed in `Approach: Defining Per Reason Controls`.
The idea is to support use case from [this](https://github.com/kubernetes-sigs/karpenter/discussions/1079#discussioncomment-8719765) discussion.
In short:
- by default -> do not tolerate any disruption (the default one)
- in specified window -> allow 25% disruption due to drift reason
Use case: we want to keep our infrastructure intact, with an exception when we want to update the fleet during the maintenance window (which happens at weekends).

Reply:
```yaml
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: default
spec: # This is not a complete NodePool Spec.
  disruption:
    defaults:
      budgets: 
        - nodes: 0% 
          schedule: "* * * * *"
    budgets:
    - schedule: "0 17 * * sat"
      reasons: [drifted, underutilized]
      nodes: 25%
```

Reply:
So you'd want to explicitly define the "If no other budgets are active, use this budget" type of behavior? This sounds like it'd be better specified through a different feature request. Are you willing to open up an issue for this @phajduk ?

Reply:
@phajduk that should be a separate RFC. I have been thinking about that kind of behavior but it doesn't belong in the per reason controls design.